nodes/hybrid-consensus/src/service.rs

```diff
@@ -0,0 +1,247 @@
+
+		let builder = sc_service::ServiceBuilder::new_full::<
+			runtime::opaque::Block, runtime::RuntimeApi, crate::service::Executor
+		>($config)?
+			.with_select_chain(|_config, backend| {
+				Ok(sc_client::LongestChain::new(backend.clone()))
+			})?
+			.with_transaction_pool(|config, client, _fetcher| {
+				let pool_api = sc_transaction_pool::FullChainApi::new(client.clone());
+				Ok(sc_transaction_pool::BasicPool::new(config, std::sync::Arc::new(pool_api)))
+			})?
+			.with_import_queue(|_config, client, mut select_chain, _transaction_pool| {
+				let select_chain = select_chain.take()
+					.ok_or_else(|| sc_service::Error::SelectChainRequired)?;
+				let (grandpa_block_import, grandpa_link) =
+					sc_finality_grandpa::block_import(
+						client.clone(), &(client.clone() as std::sync::Arc<_>), select_chain.clone()
+					)?;
+
+				//TODO do I need this justification import? Seems to work without it.
+				// let justification_import = grandpa_block_import.clone();
+
+				let pow_block_import = sc_consensus_pow::PowBlockImport::new(
+					grandpa_block_import,
+					client.clone(),
+					sha3pow::MinimalSha3Algorithm,
+					0, // check inherents starting at block 0
+					Some(select_chain),
+					inherent_data_providers.clone(),
+				);
+
+				let import_queue = sc_consensus_pow::import_queue(
+					Box::new(pow_block_import.clone()),
+					sha3pow::MinimalSha3Algorithm,
+					inherent_data_providers.clone(),
+				)?;
+
+				import_setup = Some((pow_block_import, grandpa_link));
+
+				Ok(import_queue)
+			})?;
+
+		(builder, import_setup, inherent_data_providers)
+	}}
+}
+
+/// Builds a new service for a full client.
+pub fn new_full(config: Configuration)
+	-> Result<impl AbstractService, ServiceError>
+{
+	let role = config.role.clone();
+	// let force_authoring = config.force_authoring;
+	let name = config.network.node_name.clone();
+	let disable_grandpa = config.disable_grandpa;
+
+	let (builder, mut import_setup, inherent_data_providers) = new_full_start!(config);
+
+	let (block_import, grandpa_link) =
+		import_setup.take()
+			.expect("Link Half and Block Import are present for Full Services or setup failed before. qed");
+
+	let service = builder
+		.with_finality_proof_provider(|client, backend| {
+			let provider = client as Arc<dyn StorageAndProofProvider<_, _>>;
+			Ok(Arc::new(GrandpaFinalityProofProvider::new(backend, provider)) as _)
+		})?
+		.build()?;
+
+	if role.is_authority() {
+		let proposer = sc_basic_authorship::ProposerFactory::new(
+			service.client(),
+			service.transaction_pool(),
+		);
+
+		let client = service.client();
+		let select_chain = service.select_chain()
+			.ok_or(ServiceError::SelectChainRequired)?;
+
+		let can_author_with =
+				sp_consensus::CanAuthorWithNativeVersion::new(client.executor().clone());
+
+		sc_consensus_pow::start_mine(
+			Box::new(block_import),
+			client,
+			MinimalSha3Algorithm,
+			proposer,
+			None, // TODO Do I need some grandpa preruntime digests?
```

Comment:
In principle the PoW doesn't know anything about GRANDPA right?

Reply:
That's right. My understanding is that this is a way to tell the PoW algorithm, "Hey I need you to include this preruntime digest. I don't need you to understand what it means or why it's necessary, I just need you to put it in the blocks you author." If grandpa is expecting any preruntime digests, I believe they belong here.

That being said it seems grandpa is not expecting any pre-runtime digests based on the fact that this node seems to run correctly when I supply `None`. @andresilva ca you confirm that grandpa does not expect preruntime digests?

Also a general question that I've never understood. What is the difference (in intent or semantics) between a preruntime digest and an inherent?

Reply:
GRANDPA is not expecting any preruntime digests. Currently this wouldn't be easy to implement generically since the authorship pipeline isn't really pluggable (i.e. the preruntime digest would need to be added by BABE but it would have no way to fetch digests from some other source).

> Also a general question that I've never understood. What is the difference (in intent or semantics) between a preruntime digest and an inherent?

I am also not sure about this maybe we can ask in one of the channels? I guess for inherent data we usually have a generic way of providing it? (i.e. timestamp data -> we have a clock)